src/cluster_legacy.c

```diff
@@ -4939,6 +4939,8 @@ void clusterPropagatePublish(robj *channel, robj *message, int sharded) {
     while ((node = clusterNodeIterNext(&iter)) != NULL) {
         if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
         if (nodeSupportsLightMsgHdrForPubSub(node)) {
+            /* Avoid sending pub/sub light weight message until the bidirectional link(s) have been established. */
+            if (!node->link || node->pong_received < node->link->ctime) continue;
```

Comment:
Good finding!

The bug was introduced when we added the light message, so I'm thinking we should fall back to the heavy-weight message instead of just skipping the message propagation.

```suggestion
        if (nodeSupportsLightMsgHdrForPubSub(node) && node->link && node->pong_received >= node->link->ctime) {
            /* Avoid sending pub/sub light weight message until the bidirectional link(s) have been established. */
```

Maybe even better is to do the check for `node->pong_received >= node->link->ctime` inside the definition of `nodeSupportsLightMsgHdrForPubSub()`?

Reply:
We can turn it into a static-inline function. (It's better than macros because it doesn't evaluate the argument multiple times and it's type checked and more readable.) We could also move it to `cluster_legacy.c` which is the only place where it's used.

```c
static inline bool nodeSupportsLightMsgHdrForPubSub(clusterNode *n) {
    return n->link && n->pong_received >= n->link->ctime &&
           (n->flags & CLUSTER_NODE_LIGHT_HDR_PUBLISH_SUPPORTED);
}
static inline bool nodeSupportsLightMsgHdrForModule(clusterNode *n) {
    return n->link && n->pong_received >= n->link->ctime &&
           (n->flags & CLUSTER_NODE_LIGHT_HDR_MODULE_SUPPORTED);
}
```

Reply:
Thank you! Did exactly that!